Firestore/Source/Local/FSTLocalStore.h

```diff
@@ -163,7 +166,7 @@ NS_ASSUME_NONNULL_BEGIN
 - (void)releaseQuery:(FSTQuery *)query;
 
 /** Runs @a query against all the documents in the local store and returns the results. */
-- (FSTDocumentDictionary *)executeQuery:(FSTQuery *)query;
+- (firebase::firestore::model::MaybeDocumentMap)executeQuery:(FSTQuery *)query;
```

Comment:
This is one notable change: returning a map of `FSTMaybeDocument`s instead of `FSTDocument`s. It's because unlike Objective-C, `SortedMap<K, Derived*>` cannot be converted to `SortedMap<K, Base*>`, and I can't think of a good workaround. While returning `FSTDocument`s could potentially be more efficient by saving a cast, in practice I don't think this is ever used -- all the maps of documents are eventually passed to `FSTView computeChangesWithDocuments`, which takes a map of maybe documents.

Reply:
You could add a conversion function. It'd be O(n), but presuming you're going to do something with (all) the documents, you're (at least) O(n) anyways.

As it currently stands, this PR assumes elsewhere that these MaybeDocs are actual Docs. The assumption's currently correct, but feels a bit dangerous. If I'm the next one to touch this executeQuery function, I'd happily add the ability to return other subclasses of MaybeDoc because the API says that's fine. You could fix that with documentation, but then I'd have to actually read it. :/

Reply:
I don't want to impose the performance penalty from building a tree of shared pointers (in the implementation of `SortedMap`) for potentially thousands of documents, especially considering that this PR is partially motivated by optimization.

On one hand, I'm not sure it makes sense for `executeQuery` to return `FSTDeletedDocument`s or `FSTUnknownDocument`s. On the other hand, I totally agree that this would be best reflected in the return type.

Most of the trouble comes from functions which accept a map of `FSTMaybeDocument`s. Passing iterators doesn't work, because the functions may want to use more map functionality than just iterating over it. One workaround would be to make those functions template functions. The other is to have two overloads, one taking `MaybeDocumentMap` and the other `DocumentMap` (in Objective-C I presume they would have to have different names), and try to come up with a mostly-shared implementation. What do you think?

Reply:
Ok, so I took a stab at it. AFAICT, Objective-C supports neither overloading nor templates, so what I ended up doing is:
* have two versions of both `computeView` functions, one taking a `DocumentMap` and the other taking a `MaybeDocumentMap`;
* in the implementation file, have both of them delegate to a template function;
* the only difference in the template function is when accessing values from the map. I added an overloaded function which either casts the given `FSTMaybeDocument` to an `FSTDocument`, or just returns the given `FSTDocument`. In theory, this is more efficient because when dealing with a `DocumentMap`, it avoids casting and type checking altogether.

What do you think?

Reply:
I like it. If you do too, then lets proceed with this version.

Reply:
I certainly think it's an improvement, thanks for raising this up.